dsl/read.go

```diff
@@ -83,64 +98,89 @@ func ReadHopsFileContents(hopsFileContent []FileContent) (*hcl.BodyContent, stri
-
-	return files, nil
-}
-
-// readHopsDir retrieves the content of all .hops files in a directory,
-// including sub directories, and returns then as a slice of fileContents
-func readHopsDir(dirPath string) ([]FileContent, error) {
-	filePaths := []string{}
+	err := filepath.WalkDir(root, func(path string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
 
-	err := filepath.WalkDir(dirPath, func(path string, d fs.DirEntry, err error) error {
+		// Get relative path from the root
+		relativePath, err := filepath.Rel(root, path)
 		if err != nil {
 			return err
 		}
-		// Exclude directories whose name starts with '..'
-		// This is because kubernetes configMaps create a set of symlinked
-		// directories for the mapped files and we don't want to pick those
-		// up. Those directories are named '..<various names>'
-		// Example:
-		// /my-config-map-dir
-		// |-- my-key -> ..data/my-key
-		// |-- ..data -> ..2023_10_19_12_34_56.789012345
-		// |-- ..2023_10_19_12_34_56.789012345
-		// |   |-- my-key
-		if d.IsDir() && strings.HasPrefix(d.Name(), "..") {
-			return filepath.SkipDir
+
+		if d.IsDir() {
+			// Exclude directories whose name starts with '..'
+			// This is because kubernetes configMaps create a set of symlinked
+			// directories for the mapped files and we don't want to pick those
+			// up. Those directories are named '..<various names>'
+			// Example:
+			// /my-config-map-dir
+			// |-- my-key -> ..data/my-key
+			// |-- ..data -> ..2023_10_19_12_34_56.789012345
+			// |-- ..2023_10_19_12_34_56.789012345
+			// |   |-- my-key
+			if strings.HasPrefix(d.Name(), "..") {
+				return filepath.SkipDir
+			}
+
+			// Skip any second children of the root (i.e. root/sub, yes, root/sub/sub, no)
+			if strings.Count(relativePath, string(filepath.Separator)) > 1 {
+				return filepath.SkipDir
+			}
+
+			return nil
 		}
-		if !d.IsDir() && filepath.Ext(path) == ".hops" {
-			filePaths = append(filePaths, path)
+
+		// Files in root (i.e root/a.hops), and anything other than first
+		// child directory of the root (i.e. root/sub/sub/a.hops) are skipped
+		if strings.Count(relativePath, string(filepath.Separator)) != 1 {
+			return nil
 		}
+
+		// Ensure only 1 .hops file per directory
+		if filepath.Ext(path) == HopsExt {
+			dirOnly := filepath.Dir(relativePath)
+			if seenPath[dirOnly] {
+				return fmt.Errorf("Only one hops file per directory allowed: %s", relativePath)
+			}
+
+			seenPath[dirOnly] = true
+		}
+
+		// Add file to list (both .hops and other files)
+		filePaths = append(filePaths, path)
+
 		return nil
 	})
+	// File walking is over, check for errors
 	if err != nil {
 		return nil, err
 	}
 
 	// Sort the file paths to ensure consistent order
 	sort.Strings(filePaths)
 
+	return filePaths, nil
+}
+
+// readHops retrieves the content of all .hops and other files in all the
+// first child subdirectories of dirPath (excluding dirs with '..' prefix)
+// and returns them as a slice of fileContents
```

Comment:
Picky comment on my part here, but standard practice for go docs is the first part explains the function (without internal behaviour or caveats) and then there's a blank comment line, followed by any deeper comment explaining things like the exclusion of `..` etc.

This is due to how the go pkg docs render these for you. Makes for much nicer reading

Reply:
Fixed